File: src/mapper.py
"""Content mapper to convert markdown files to WordPress format."""

from typing import Dict, Any, List, Optional
from .types import MarkdownFile, WordPressPost, Config
from .wordpress_client import WordPressClient


class ContentMapper:
    """Maps markdown content to WordPress post format."""
    
    def __init__(self, config: Config, wp_client: WordPressClient):
        self.config = config
        self.wp_client = wp_client
    
    def map_to_wordpress(self, md_file: MarkdownFile) -> Dict[str, Any]:
        """Convert a markdown file to WordPress post data."""
        fm = md_file.front_matter
        
        # Basic post data
        post_data = {
            "title": fm.title,
            "content": md_file.html or md_file.content,
            "slug": fm.slug,
            "status": fm.status or self.config.default_status,
            "excerpt": fm.subtitle or fm.description or "",
        }
        
        # Handle dates
        if fm.date_published:
            post_data["date"] = fm.date_published
        if fm.date_updated:
            post_data["modified"] = fm.date_updated
        
        # Handle author
        if fm.authors and len(fm.authors) > 0:
            author_id = self._resolve_author(fm.authors[0])
            if author_id:
                post_data["author"] = author_id
        
        # Handle taxonomies
        if fm.tags:
            tag_ids = self._resolve_taxonomy_terms("tags", fm.tags)
            if tag_ids:
                post_data["tags"] = tag_ids
        
        if fm.categories:
            cat_ids = self._resolve_taxonomy_terms("categories", fm.categories)
            if cat_ids:
                post_data["categories"] = cat_ids
        
        # Handle meta fields
        meta = {}
        
        # Always set external ID for idempotency
        meta["_external_id"] = fm.external_id or fm.slug
        
        # Type-specific meta
        if fm.type == "event":
            if fm.start_date:
                meta["start_date"] = fm.start_date
            if fm.end_date:
                meta["end_date"] = fm.end_date
            if fm.location_name:
                meta["location_name"] = fm.location_name
            if fm.location_address:
                meta["location_address"] = fm.location_address
            if fm.registration_url:
                meta["registration_url"] = fm.registration_url
            if fm.host:
                meta["host"] = fm.host if isinstance(fm.host, str) else ", ".join(fm.host)
        
        elif fm.type == "podcast":
            if fm.episode_number:
                meta["episode_number"] = fm.episode_number
            if fm.audio_url:
                meta["audio_url"] = fm.audio_url
            if fm.duration:
                meta["duration"] = fm.duration
            if fm.guests:
                meta["guests"] = ", ".join(fm.guests)
            if fm.show:
                meta["show"] = fm.show
        
        elif fm.type == "blog" and fm.featured is not None:
            meta["featured"] = fm.featured
        
        if meta:
            post_data["meta"] = meta
        
        return post_data
    
    def _resolve_author(self, author: str) -> Optional[int]:
        """Resolve author name/email to WordPress user ID."""
        # Try email first
        if "@" in author:
            user_id = self.wp_client.find_user_by_email(author)
            if user_id:
                return user_id
        
        # Try display name
        user_id = self.wp_client.find_user_by_name(author)
        if user_id:
            return user_id
        
        # Use default author
        return self.wp_client.find_user_by_name(self.config.default_author)
    
    def _resolve_taxonomy_terms(self, taxonomy: str, terms: List[str]) -> List[int]:
        """Resolve taxonomy term names to IDs."""
        term_ids = []
        
        for term in terms:
            term_id = self.wp_client.get_or_create_taxonomy_term(taxonomy, term)
            if term_id:
                term_ids.append(term_id)
        
        return term_ids
    
    def get_post_type(self, content_type: str) -> str:
        """Get WordPress post type for content type."""
        type_mapping = {
            "blog": "posts",
            "news": "posts", 
            "page": "pages",
            "event": "event",  # Assumes custom post type
            "podcast": "podcast"  # Assumes custom post type
        }
        
        return type_mapping.get(content_type, "posts")

File: src/core/mapper.ts
import { MarkdownFile, WordPressPost, Config, Mapping, FrontMatter } from '../types';
import { WordPressClient } from '../wordpress/client';
import { MediaHandler } from '../wordpress/media';
import logger from '../utils/logger';
import path from 'path';

export class ContentMapper {
  private wpClient: WordPressClient;
  private mediaHandler: MediaHandler;
  private config: Config;
  private mapping: Mapping;
  
  constructor(config: Config, mapping: Mapping) {
    this.config = config;
    this.mapping = mapping;
    this.wpClient = new WordPressClient(config);
    this.mediaHandler = new MediaHandler(config);
  }
  
  async mapToWordPress(file: MarkdownFile): Promise<WordPressPost> {
    const type = file.frontMatter.type || 'blog';
    const typeMapping = this.mapping.post_types[type];
    
    if (!typeMapping) {
      throw new Error(`No mapping found for type: ${type}`);
    }
    
    // Validate required fields
    this.validateRequiredFields(file.frontMatter, typeMapping.required_fields);
    
    // Create base post object
    const post: WordPressPost = {
      title: file.frontMatter.title,
      slug: file.frontMatter.slug!,
      status: file.frontMatter.status || this.config.defaults.status,
      content: '',
      meta: {}
    };
    
    // Map fields
    if (typeMapping.fields) {
      for (const [wpField, sourceField] of Object.entries(typeMapping.fields)) {
        if (sourceField === 'BODY') {
          // Process content with media handler
          const basePath = path.dirname(file.path);
          post.content = await this.mediaHandler.processContentImages(file.html!, basePath);
        } else if (file.frontMatter[sourceField as keyof FrontMatter] !== undefined) {
          const value = file.frontMatter[sourceField as keyof FrontMatter];
          this.setNestedProperty(post, wpField, value);
        }
      }
    }
    
    // Map excerpt
    if (file.frontMatter.subtitle || file.frontMatter.description) {
      post.excerpt = file.frontMatter.subtitle || file.frontMatter.description;
    }
    
    // Map dates
    if (file.frontMatter.date_published) {
      post.date = file.frontMatter.date_published;
      post.date_gmt = this.toGMT(file.frontMatter.date_published);
    }
    
    if (file.frontMatter.date_updated) {
      post.modified = file.frontMatter.date_updated;
      post.modified_gmt = this.toGMT(file.frontMatter.date_updated);
    }
    
    // Handle featured image
    if (file.frontMatter.featured_image && this.config.defaults.media.upload) {
      const mediaId = await this.mediaHandler.uploadFeaturedImage(
        file.frontMatter.featured_image,
        file.frontMatter.title
      );
      if (mediaId) {
        post.featured_media = mediaId;
      }
    }
    
    // Handle author
    const authorId = await this.resolveAuthor(file.frontMatter.authors);
    if (authorId) {
      post.author = authorId;
    }
    
    // Handle taxonomies
    if (typeMapping.taxonomies) {
      await this.mapTaxonomies(post, file.frontMatter, typeMapping.taxonomies);
    }
    
    // Map meta fields
    if (typeMapping.meta) {
      for (const [metaKey, sourceField] of Object.entries(typeMapping.meta)) {
        const value = file.frontMatter[sourceField as keyof FrontMatter];
        if (value !== undefined) {
          post.meta![metaKey] = value;
        }
      }
    }
    
    // Always set external_id for idempotency
    post.meta!._external_id = file.frontMatter.external_id || file.frontMatter.slug;
    
    return post;
  }
  
  private validateRequiredFields(frontMatter: FrontMatter, requiredFields: string[]): void {
    const missing: string[] = [];
    
    for (const field of requiredFields) {
      if (!frontMatter[field as keyof FrontMatter]) {
        missing.push(field);
      }
    }
    
    if (missing.length > 0) {
      throw new Error(`Missing required fields: ${missing.join(', ')}`);
    }
  }
  
  private setNestedProperty(obj: any, path: string, value: any): void {
    const parts = path.split('.');
    let current = obj;
    
    for (let i = 0; i < parts.length - 1; i++) {
      if (!current[parts[i]]) {
        current[parts[i]] = {};
      }
      current = current[parts[i]];
    }
    
    current[parts[parts.length - 1]] = value;
  }
  
  private toGMT(dateStr: string): string {
    const date = new Date(dateStr);
    return date.toISOString();
  }
  
  private async resolveAuthor(authors?: string[]): Promise<number | null> {
    if (!authors || authors.length === 0) {
      return null;
    }
    
    const firstAuthor = authors[0];
    
    // Check if it's an email
    if (firstAuthor.includes('@')) {
      const userId = await this.wpClient.getUserIdByEmail(firstAuthor);
      if (userId) return userId;
    }
    
    // Try by display name
    const userId = await this.wpClient.getUserIdByName(firstAuthor);
    if (userId) return userId;
    
    // Use fallback author
    const fallbackId = await this.wpClient.getUserIdByName(this.config.defaults.author_fallback);
    
    if (fallbackId) {
      logger.debug(`Using fallback author for: ${firstAuthor}`);
      return fallbackId;
    }
    
    logger.warn(`Could not resolve author: ${firstAuthor}`);
    return null;
  }
  
  private async mapTaxonomies(
    post: WordPressPost,
    frontMatter: FrontMatter,
    taxonomyMapping: { [key: string]: string }
  ): Promise<void> {
    for (const [wpTaxonomy, sourceField] of Object.entries(taxonomyMapping)) {
      const terms = frontMatter[sourceField as keyof FrontMatter] as string[] | undefined;
      
      if (!terms || !Array.isArray(terms) || terms.length === 0) {
        continue;
      }
      
      const termIds: number[] = [];
      
      for (const term of terms) {
        const termId = await this.wpClient.getTaxonomyTermId(wpTaxonomy, term);
        if (termId) {
          termIds.push(termId);
        }
      }
      
      if (termIds.length > 0) {
        // Map to correct post field based on taxonomy type
        if (wpTaxonomy === 'post_tag' || wpTaxonomy === 'tags') {
          post.tags = termIds;
        } else if (wpTaxonomy === 'category' || wpTaxonomy === 'categories') {
          post.categories = termIds;
        } else {
          // Custom taxonomy - store in meta or use REST API field name
          if (!post.meta) post.meta = {};
          post.meta[wpTaxonomy] = termIds;
        }
      }
    }
  }
  
  getPostType(contentType: string): string {
    const typeMapping = this.mapping.post_types[contentType];
    
    if (!typeMapping || !typeMapping.endpoint) {
      return 'posts'; // Default to posts
    }
    
    // Extract post type from endpoint
    const match = typeMapping.endpoint.match(/\/wp-json\/wp\/v2\/(\w+)/);
    
    if (match) {
      return match[1];
    }
    
    return 'posts';
  }
  
  async testConnection(): Promise<boolean> {
    return await this.wpClient.testConnection();
  }
}